storage-proofs/porep/src/nse/circuit/hash.rs

```diff
@@ -0,0 +1,276 @@
+use bellperson::{
+    gadgets::{boolean::Boolean, num, sha256::sha256 as sha256_circuit, uint32::UInt32},
+    ConstraintSystem, SynthesisError,
+};
+use ff::PrimeField;
+use paired::bls12_381::{Bls12, Fr};
+use storage_proofs_core::{
+    gadgets::{constraint, encode, multipack, uint64::UInt64},
+    util::reverse_bit_numbering,
+};
+
+use super::super::Config;
+
+pub fn derive_first_layer_leaf<CS: ConstraintSystem<Bls12>>(
+    mut cs: CS,
+    replica_id: &[Boolean],
+    challenge_index_num: &UInt64,
+    layer: u32,
+) -> Result<num::AllocatedNum<Bls12>, SynthesisError> {
+    // TODO: can we reuse this allocation accross challenges?
+    let layer_index = UInt32::alloc(cs.namespace(|| "layer_index"), Some(layer))?;
+
+    // prefix
+    let mut ciphertexts = hash_prefix(challenge_index_num, layer_index)?;
```

Comment:
Is the word 'ciphertext' being used here for any reason other than because it has gradually leaked into the code base over time? I think it was originally used for the output of Sloth, which later became just XOR. In that context, it still made sense because a key was being used to transform node values.

By SDR, this was no longer really what was happening, but I think the name stuck around. Now it seems like we're using it for hash input.

I've recapitulated a probably poor history as a guess about how this naming came to be. If there's a different reason for it, please let me know. Otherwise, I propose we adopt new and less confusing naming moving forward.